crates/wasmparser/src/binary_reader.rs

```diff
@@ -242,78 +300,458 @@ impl<'a> BinaryReader<'a> {
+                        "only aliases to outer types are supported in component type definitions",
+                        offset,
+                    )),
+                }
+            }
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!(
+                        "invalid leading byte (0x{:x}) for component type definition",
+                        x
+                    ),
+                    self.original_position() - 1,
+                ))
+            }
         })
     }
 
     pub(crate) fn read_instance_type(&mut self) -> Result<InstanceType<'a>> {
-        Ok(InstanceType {
-            exports: self.read_export_types()?,
+        Ok(match self.read_u8()? {
+            0x01 => InstanceType::Type(self.read_component_type_def()?),
+            0x07 => InstanceType::Export {
+                name: self.read_string()?,
+                ty: self.read_var_u32()?,
+            },
+            0x09 => {
+                let offset = self.original_position();
+                let alias = self.read_alias()?;
+                match alias {
+                    Alias::OuterType { count, index } => InstanceType::OuterType { count, index },
+                    _ => return Err(BinaryReaderError::new(
+                        "only aliases to outer types are supported in instance type definitions",
+                        offset,
+                    )),
+                }
+            }
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!(
+                        "invalid leading byte (0x{:x}) for instance type definition",
+                        x
+                    ),
+                    self.original_position() - 1,
+                ))
+            }
         })
     }
 
-    fn read_export_types(&mut self) -> Result<Box<[ExportType<'a>]>> {
-        let pos = self.original_position();
-        let exports_len = self.read_var_u32()? as usize;
-        if exports_len > MAX_WASM_EXPORTS {
-            return Err(BinaryReaderError::new("exports size is out of bound", pos));
-        }
-        (0..exports_len).map(|_| self.read_export_type()).collect()
+    pub(crate) fn read_interface_type(&mut self) -> Result<InterfaceType> {
+        let position = self.position;
+        Ok(match self.read_u8()? {
+            0x7f => InterfaceType::Unit,
+            0x7e => InterfaceType::Bool,
+            0x7d => InterfaceType::S8,
+            0x7c => InterfaceType::U8,
+            0x7b => InterfaceType::S16,
+            0x7a => InterfaceType::U16,
+            0x79 => InterfaceType::S32,
+            0x78 => InterfaceType::U32,
+            0x77 => InterfaceType::S64,
+            0x76 => InterfaceType::U64,
+            0x75 => InterfaceType::F32,
+            0x74 => InterfaceType::F64,
+            0x73 => InterfaceType::Char,
+            0x72 => InterfaceType::String,
+            x if x >= 0x80 => {
```

Comment:
This is indeed incorrect here and in `wasm-encoder`. The spec I think is calling for `SLEB128` encoding of the compound type index (`wasm-encoder` incorrectly is using `encoders::u32` and not `encoders::s32`). There should be a call to `read_var_i32` for the index here.

With SLEB128, we still have to read the first byte and unwind if it's `>= 0x80`, not unlike what we do in `read_blocktype` for core wasm.

Reply:
At the bare minimum this could also be updated to peek at the current byte (after refactoring `read_blocktype` too) and if it's not one of the expected values, delegate to `read_var_i32` for a compound type index.